### PR TITLE
feat(mcp): query tools — get_callers/callees/deps, find_path, search_code (T5/T7/T8)

### LINKED ISSUE #653: [MCP T5] get_callers / get_callees / get_dependencies tools
Phase 1 ticket T5. Depends on #652 (T4 index_repo).

## Context

Three closely related neighbor-traversal tools that share a single helper. They are the bread-and-butter of agent-driven code navigation: "what calls this", "what does this call", "what does this depend on". The underlying graph operation already exists — `AsyncGraphQuery.get_neighbors()` (api/graph.py:690-712) — and the existing CLI command `cgraph neighbors` (api/cli.py:288-311) already exposes it. This ticket wraps that single operation as three differently-shaped MCP tools.

## Scope

In:
- `api/mcp/tools/structural.py` adds three tools sharing one helper:
  - `get_callers(symbol_id, limit=50)` → `rel="CALLS", dir="IN"`
  - `get_callees(symbol_id, limit=50)` → `rel="CALLS", dir="OUT"`
  - `get_dependencies(symbol_id, rels=None, limit=50)` → accepts a list of relation types (default `["IMPORTS", "CALLS", "DEFINES"]`) and traverses any.
- All three return typed node lists with `id`, `name`, `label`, `file`, `line`, `relation` (the rel that connected them), `direction` (IN/OUT).
- Wraps `AsyncGraphQuery.get_neighbors()` (api/graph.py:690-712). Existing helpers `Graph.function_called_by()` (api/graph.py:383-391) and `Graph.function_calls()` (api/graph.py:373-381) are also valid implementation choices — use whichever produces the cleanest call site.
- Tests in `tests/mcp/test_neighbors.py`:
  - **Unit (mocked graph)**: empty result, single result, multi-rel filter for `get_dependencies`.
  - **Integration**: against the T3 fixture's expected callers/callees from `expected.yaml`. For `service` in the fixture: `get_callers(service)` returns `entrypoint`; `get_callees(service)` returns `repo`.
  - **CLI parity**: assert that `cgraph neighbors --rel CALLS --dir in <id>` returns the same nodes as the MCP `get_callers` tool for the same input. Same for `--dir out` ↔ `get_callees`.
  - **Protocol round-trip** for each of the three tools.

Out:
- Variable-depth traversal (T6 `impact_analysis` does that).
- Path discovery (T7 `find_path`).

## Files to create / modify

- `api/mcp/tools/structural.py` — three tool registrations + one shared helper
- modified `api/mcp/server.py` if needed (auto-discovery of new tools)
- new `tests/mcp/test_neighbors.py`

## Acceptance criteria

- [ ] All three tools registered with FastMCP and discoverable via `session.list_tools()`.
- [ ] Each tool's input schema is correct: `get_callers` and `get_callees` take `symbol_id: str, limit: int = 50`; `get_dependencies` adds `rels: list[str] | None = None`.
- [ ] Unit tests cover empty result, single result, and multi-rel filter — all with mocked graph.
- [ ] Integration test against T3 fixture: for at least one named symbol in `expected.yaml`, asserts the callers and callees match the contract.
- [ ] CLI parity test asserts MCP tool output matches `cgraph neighbors` output for the same input — catches drift.
- [ ] Protocol round-trip test for each of the three tools.
- [ ] CI workflow #649 green on all of the above.
- [ ] Each tool honors `limit` and never returns more than the requested count.

## Dependencies

- Blocks: nothing in Phase 1 (sibling to T6/T7/T8).
- Blocked by: #652 (T4 — needs the graph populated by `index_repo`).

## Out of scope (do NOT do in this PR)

- `impact_analysis` (T6 — different ticket; variable-depth traversal).
- `find_path` (T7).
- `search_code` (T8).
- Cross-branch queries.

## Notes for the implementer

- Three tools sharing one helper means: don't write the Cypher three times. The helper is the only place that touches `AsyncGraphQuery.get_neighbors()`.
- `get_dependencies` defaults to `["IMPORTS", "CALLS", "DEFINES"]` because that's the most useful "what does this depend on" set. Document the default in the tool docstring (which becomes the MCP schema description).
- Each result row should include `relation` and `direction` so the agent can disambiguate (e.g. when calling `get_dependencies` with multiple rels, the agent needs to know which rel each result came from).
- Reuse the `_run_protocol_test` helper from T4.
- Reference: `api/cli.py:288-311` shows exactly how the CLI calls `get_neighbors`. Mirror that behavior.

### PR DESCRIPTION
**Pull request by DvirDukhan**
## Prerequisites (merge order)

Merge in order — this PR is stacked on:
1. #678 — `index_repo` tool (T4)

Base: #678.

---

Bundles **T5 (#653)**, **T7 (#655)**, and **T8 (#656)**. Stacked on #678 → #677 → #676 → #675 → #666.

## Tools

| Tool | Direction | Backing op |
|---|---|---|
| `search_code(prefix)` | — | `prefix_search` (FalkorDB fulltext) |
| `get_callers(symbol_id)` | IN | inline Cypher `(src)-[:CALLS]->(n)` |
| `get_callees(symbol_id)` | OUT | inline Cypher `(n)-[:CALLS]->(dest)` |
| `get_dependencies(symbol_id, rels=[CALLS, IMPORTS, DEFINES])` | OUT | multi-rel + dedup |
| `find_path(source_id, dest_id, max_paths=5)` | — | `find_paths` (CALLS-only) |

Shared helpers in `api/mcp/tools/structural.py`: `_node_summary` (flattens `encode_node`'s `{id, labels, properties: {...}}` and drops the `Searchable` label), `_coerce_node_id` (int or stringified int; rejects bool), `_project_arg` (returns `AsyncGraphQuery`).

## Tests

`tests/mcp/test_query_tools.py` — 13 tests, all green against FalkorDB on 6390. Full MCP suite: 25 passed in 23s.

## Side fix

Dropped stray `tests/mcp/fixtures/sample_project/venv/` that was polluting prefix-search results.

## Out of scope

T6 (#654, `impact_analysis`) — variable-depth Cypher with cycle avoidance, deserves its own PR.

Closes #653, #655, #656.

## Summary by CodeRabbit

* **New Features**
  * Added optional result limits to code search and path-finding operations for improved control and performance
  * Introduced new code navigation and analysis tools: symbol search with prefix matching, caller/callee tracking, multi-relation dependency aggregation, and call-path discovery between code symbols

* **Tests**
  * Added comprehensive test suite for code navigation and query capabilities

### CODE REVIEW

File: api/mcp/tools/structural.py [hunk 1]
```diff
@@ -182,3 +182,242 @@ def _payload(project) -> dict[str, Any]:
...
+    # [node, edge, node, edge, ..., node] list; we strip edges and surface
+    # only the node sequence — that's what agents typically want.
+    paths: list[dict[str, Any]] = []
+    for entry in raw[:max_paths]:
+        node_seq = [
+            _node_summary(x)
+            for x in entry
+            # Edges in the alternating list carry a top-level ``relation``
+            # key (from ``encode_edge``); nodes carry ``properties``.
+            if isinstance(x, dict) and "properties" in x
```
Comment (galshubeli) [inline api/mcp/tools/structural.py]: The premise (per the comment) is "nodes carry properties, edges carry relation." That's wrong. I checked FalkorDB's Edge class — edge.py has self.properties = properties or {}, so encode_edge → vars(e) produces a dict that also
  contains "properties". AsyncGraphQuery.find_paths returns an alternating [node, edge, node, …] list, so this filter keeps every element. Each edge then goes through _node_summary and comes out as a bogus {id: <edge_id>, name: 
  null, label: null, file: null, line: null} entry interleaved between the real nodes

Comment (galshubeli) [reply api/mcp/tools/structural.py]: the test doesn't catch it (test_find_path_entrypoint_to_db): it only asserts seq[0]["name"]=="entrypoint", seq[-1]["name"]=="db", and len(seq) >= 2. Since the alternating list always starts/ends with a node and the bogus
  edge entries sit in the middle, all three assertions pass while the payload is wrong.
  → Fix: discriminate on a node-only key — if isinstance(x, dict) and "labels" in x (edges have no labels), or exclude "relation" in x. Then add a test asserting every element in seq has a non-null name/label, or len(seq) ==
  expected_node_count.

Comment (DvirDukhan) [reply api/mcp/tools/structural.py]: Fixed in 2df968b. The filter now discriminates on `labels` (which only `encode_node` emits) instead of `properties`. FalkorDB's `Edge` also exposes a `properties` attribute, so the old filter let edges slip through as bogus all-null `_node_summary` entries — edges carry `relation`/`src_node`/`dest_node` and no `labels`. The find_path test now asserts every element in each path resolves a non-null name and label, which fails against the old `properties` filter (verified by reverting).

File: api/mcp/tools/structural.py [hunk 2]
```diff
@@ -208,3 +208,242 @@ def _payload(project) -> dict[str, Any]:
...
+    rels: Optional[list[str]] = None,
+    limit: int = 50,
+) -> list[dict[str, Any]]:
+    if rels is None:
+        rels = ["IMPORTS", "CALLS", "DEFINES"]
+    # Aggregate across relations; preserve ordering and dedupe by id.
+    seen: set[Any] = set()
+    out: list[dict[str, Any]] = []
+    for rel in rels:
+        rows = await _neighbors_payload(project, branch, symbol_id, rel, "OUT", limit)
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply api/mcp/tools/structural.py]: Fixed in b590259. Relation types are now validated with `_validate_relation` (must match `^[A-Za-z_][A-Za-z0-9_]*$`) before being interpolated into the Cypher `-[e:{rel}]->` — FalkorDB can't bind relationship types as parameters, so this closes the injection vector. Added a regression test (`test_get_dependencies_rejects_injected_relation`). Also fixed the over-fetch: each relation now requests only the remaining `limit - len(out)`, so total work is bounded by `limit` instead of `limit * len(rels)`.

File: api/mcp/tools/structural.py [hunk 3]
```diff
@@ -208,3 +208,242 @@ def _payload(project) -> dict[str, Any]:
...
+    for entry in raw[:max_paths]:
+        node_seq = [
+            _node_summary(x)
+            for x in entry
+            # Edges in the alternating list carry a top-level ``relation``
+            # key (from ``encode_edge``); nodes carry ``properties``.
+            if isinstance(x, dict) and "properties" in x
+        ]
+        paths.append({"path": node_seq})
+    return paths
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply api/mcp/tools/structural.py:455]: Fixed in b590259. `AsyncGraphQuery.find_paths` now takes an optional `limit` and appends `LIMIT $limit` to the Cypher; `find_path` passes `max_paths` down so the DB bounds enumeration instead of returning all paths and slicing in Python. The HTTP/CLI callers pass no limit (default `None`) so their behavior is unchanged.

File: api/mcp/tools/structural.py [hunk 4]
```diff
@@ -208,3 +208,242 @@ def _payload(project) -> dict[str, Any]:
...
+    project: str,
+    branch: Optional[str] = None,
+    limit: int = 20,
+) -> list[dict[str, Any]]:
+    g = _project_arg(project, branch)
+    try:
+        raw = await g.prefix_search(prefix)
+    finally:
+        await g.close()
+    return [_node_summary(node) for node in raw[:limit]]
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply api/mcp/tools/structural.py]: Fixed in b590259. `prefix_search` now parameterizes `LIMIT $limit` (default 10 preserves existing HTTP/CLI behavior) and `search_code` passes its `limit` through, so the tool's `limit` is enforced in the database rather than capped at 10 then sliced.

File: api/mcp/tools/structural.py [hunk 5]
```diff
@@ -208,3 +208,242 @@ def _payload(project) -> dict[str, Any]:
...
+        "`symbol_id` is the integer node id returned by `search_code` or "
+        "other tools."
+    ),
+)
+async def get_callers(
+    symbol_id: Any,
+    project: str,
+    branch: Optional[str] = None,
+    limit: int = 50,
+) -> list[dict[str, Any]]:
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply api/mcp/tools/structural.py:349]: Fixed in b590259 — `symbol_id` is now annotated `int | str`.

File: api/mcp/tools/structural.py [hunk 6]
```diff
@@ -208,3 +208,242 @@ def _payload(project) -> dict[str, Any]:
...
+    description=(
+        "Return functions that the given symbol calls (outgoing CALLS edges)."
+    ),
+)
+async def get_callees(
+    symbol_id: Any,
+    project: str,
+    branch: Optional[str] = None,
+    limit: int = 50,
+) -> list[dict[str, Any]]:
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply api/mcp/tools/structural.py:364]: Fixed in b590259 — `symbol_id` is now annotated `int | str`.

File: api/mcp/tools/structural.py [hunk 7]
```diff
@@ -208,3 +208,242 @@ def _payload(project) -> dict[str, Any]:
...
+        "Useful for 'what does this depend on' queries."
+    ),
+)
+async def get_dependencies(
+    symbol_id: Any,
+    project: str,
+    branch: Optional[str] = None,
+    rels: Optional[list[str]] = None,
+    limit: int = 50,
+) -> list[dict[str, Any]]:
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply api/mcp/tools/structural.py:382]: Fixed in b590259 — `symbol_id` is now annotated `int | str`.

File: api/mcp/tools/structural.py [hunk 8]
```diff
@@ -208,3 +208,242 @@ def _payload(project) -> dict[str, Any]:
...
+        "empty list when no path exists."
+    ),
+)
+async def find_path(
+    source_id: Any,
+    dest_id: Any,
+    project: str,
+    branch: Optional[str] = None,
+    max_paths: int = 10,
+) -> list[dict[str, Any]]:
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply api/mcp/tools/structural.py:427]: Fixed in b590259 — `source_id`/`dest_id` are now annotated `int | str`.

**Review by galshubeli — DISMISSED** [review]

**Review by galshubeli — APPROVED** [review]
